dtt params to filter table + general config updates

asyates opened the pull request:
Afew things added/changed. I still want to make some further adjustments (listed below), but if there is any feedback on these changes i.e. whether they're wanted or not, or if there is a better way of doing it, would be useful before merging.

**Changes:**

- dtt_minlag, dtt_width, and dtt_v, moved to filter table. I've always personally, at least for dtt_minlag and dtt_width, adjusted the code to have these defined for individually for each filter as makes the most sense to me. 

- config sql table contains used_in field which is a list of steps where it is used/defined. It can be used as a filter on msnoise admin (originally was looking to do this without having it as a column in the sql table, only in defaults.csv, but was easier just including it in the config table... and just not showing it on admin).

- adjusted  some of parameters in wavelet codes (from using 'dtt' to 'wct') to make it clear they are wavelet related, e.g. wct_minlag, wct_codacycles, wct_min_nonzero. Might rename some of these variables anyway actually, as i just chose them quickly when writing the code.

**Some images of the changes**

filter table new columns:
![filter_admin1](https://github.com/user-attachments/assets/6d5e28cd-688f-4e03-af4d-c689ff3732c1)

dropdown option in config with used_in filter:
![msnoise_admin3](https://github.com/user-attachments/assets/dc141aaa-16fd-41ed-bb30-56236f7cd346)
![msnoise_admin2](https://github.com/user-attachments/assets/3763b341-0b65-4c1b-a25f-1a99afa00206)

wct param name changes:
![msnoise_admin4](https://github.com/user-attachments/assets/3f3923ef-e126-4c12-9dec-6b08e66f4423)**

**To do (related to param changes)**

- Double check documentation, probably needs some updates.
- Adjust wavelet code to also include option for static lag times. Also, some of the 'default' choices should be changed.

As said, feedback welcome re. any of this!

ThomasLecocq commented:
Super work here @asyates ! so remaining mwcs_dtt_mincoh, mwcs_dtt_maxerr etc are going to be super easy to understand! that's cool !

Still thinking of splitting the MWCS+DTT ; WCT ; STR from the filter table... be in another table, and with links to the filter table's ref.. Allowing many-to-one reference

ThomasLecocq commented:
@asyates so the stretching parameters should also go in the filters' definition ?

ThomasLecocq commented:
> maybe just need to rename i.e. rather than dtt_minlag and dtt_width.... lag_win_start, lag_win_length, as not specific to dt/t (or even just window_start and window_length)
agree with the renaming. 

Still think stretching is a "dvv" method, and should be "after filter" (if you think of the steps as a chain)...

ThomasLecocq commented:
Yep, I'm really thinking of altering the schema (could be for msnoise 3) to include a DVV table that links to one (or more) filters...

the dvv table fields could be dynamic: 
ref, filter_id, method, method_parameters

where method= ["mwcs", "wct", "stretching", new method...]

the difficulty here is to create a schema that accepts the params as such.

The way wordpress does is to store the param_name=value in different ROWS , effectively splitting the "dvv row" into:

```
1: ref = 1 ; param = filter_id, value = 1
2: ref = 1 ; param = method, value = mwcs
3 ref = 1 ; param = mwcs_freqmin, value= 0.1
...

```

chat GPT might have another idea for that :)

ThomasLecocq commented:
re: many tables: it's also a good idea, it doesn't really matter how many tables there are, and allows to easily add extra methods. Table naming : %prefix + "dvv_<method_slug>" : dvv_mwcs, dvv_stretching, dvv_wct, dvv_dtw etc...

asyates commented:
> > maybe just need to rename i.e. rather than dtt_minlag and dtt_width.... lag_win_start, lag_win_length, as not specific to dt/t (or even just window_start and window_length)
> > agree with the renaming.
> 
> Still think stretching is a "dvv" method, and should be "after filter" (if you think of the steps as a chain)...

Sure, it's definitely after the filter.... but i guess if you move stretching params into the filter table then we should also be moving all mwcs_dtt_ and maybe some wct_ params in there too.

Practically I have less reason to define these at the filter level (whereas the dtt_minlag and and dtt_width in my opinion should absolutely be defined for individual filters, and are used by all methods), but if you think logically it makes sense for them to go in I can. Personally, i might prefer to leave just the params used by all methods in the filter table for now, and later look at the option of a separate DVV table as you describe.

ThomasLecocq commented:
agree globally to keep it simple:

just an edge case : if dtt_minlag is global, and dtt_lag=dynamic is set, and you're doing SC, the stretching method should go back to minlag... but that minlag & width are also filter dependent, no ?

asyates commented:
its a good point... i am not sure there is any check in place right now to catch if its SC + dynamic, because, as you suggest, it should then use minlag. I suspect it doesn't... (will double check and fix if not).

asyates commented:
> 1 CCF calculation + 10 MWCS narrowbands + 10 STR narrowbands + 1 WCT broadband --> to compare the outputs :-)

I like this idea a lot, assuming we are comfortable that we do not expect strong differences between a narrow whiten/filter -> dvv versus broad whiten -> narrow fillter -> dvv 

Does this mean the death of the filter table?

ThomasLecocq commented:
>I like this idea a lot, assuming we are comfortable that we do not expect strong differences between a narrow whiten/filter -> dvv versus broad whiten -> narrow fillter -> dvv. I have done small testing, but certainly not exhaustive, towards verifying that this is the case.

remains an open question for me, the changes you make now will allow benchmarking :)

>This would also mean the death of the filter table i guess? (just using preprocess_lowpass/_highpass).

wrong, it's still nice to be able to have different filters too... e.g. to apply filter 1 to CC+AC+SC and filter 2 to AC+SC only? (yes, that would be an extra parameter) :-)

asyates commented:
Okay... i **think** all the core functionality is now implemented. I've documented the big changes below. Still some things left to do (also documented below), but wondering whether worth merging this request now and doing the rest as further smaller pull requests (since this is already large).

Key changes:

**Many to many relationships between different steps through association tables**
By far the biggest change: New individual tables for mwcs/stretching/wct params and mwcs_dtt/wct_dtt params. For mwcs/stretching/wct params it is now possible to define many different parameter choices and point them to one or many filters. Similarly many mwcs_dtt/wct_dtt_params can be defined and pointed to many mwcs/stretching/wct params. The relationships between different levels/steps is stored in association tables, and displayed dynamically in msnoise admin. For updating/inserting via api, a list of ids is used to update the association tables (e.g. defining which filter_ids a particular set of dvv params should point to).

When saving results, the convention follows filt_id/level1_id/level2_id.... where mwcs/stretching/wct_params are level one, and dtt params are level 2. Note, no intermediate wavelet transform results are saved (prior to dt/t) as the size of such files would be huge (i checked ;) ). So the wavelet transform + dt/t are still performed in one step for now... however its still now possible to have many to many relationship between different wavelet params and wavelet_dtt params.

Fig 1:
![image](https://github.com/user-attachments/assets/1cf434d6-d0a6-4b8b-8d03-280a2a486846)

Fig 2:
![image](https://github.com/user-attachments/assets/c0a726f7-569c-4e0d-abd8-42fe89cfbf7b)

Fig 3:
![image](https://github.com/user-attachments/assets/618134ea-796b-40dc-bf9d-ceec7f4831f3)

Fig 4:
![image](https://github.com/user-attachments/assets/095db80f-78e7-4dda-93cd-9a724c5121d0)

Fig 5.
![image](https://github.com/user-attachments/assets/4689e8c2-c1c1-4fcb-a2a1-648218e02cc5)


Key point is that we are no longer relying on single global values for various parameters involved in DVV steps.

**Toggles in filter table for CC/SC/AC**
Haven't tested extensively, but should now be able to turn off CC/SC/AC at the filter level for the different steps. It is still defined globally in the config table... as it was being used in new_jobs creation so left this... but most steps will now extract the relevant components based on the toggles in the filter table.

![image](https://github.com/user-attachments/assets/ae1b8a95-4dac-4cbd-83c7-72693dfccf42)

To do (outside of this pull request?):

- Tidy up documentation, and also update logger outputs to reflect many to many relationships (i.e. looping through different param choices).
- Some of params for stretching and wct_dtt not implemented yet (some i added because should be consistent across different methods e.g. static/dynamic choice, choice of which side of CCF etc.
- maxdt needts to be updated to maxdtt
- Maybe less important, but im sure some choices of variable names could be made more intuitive... but that's easy to change later.

I had thought about having one entry in each table already in place... but might get confusing if someone working with just api (i.e. maybe not aware it's there). So i think easiest way to will be to develop a quick_start notebook.

ThomasLecocq commented:
MASSIVE work here @asyates !!!!

Is the config max_dt or max_dtt now ? 

It's of course super hard to review such a massive change, I'll try to give it a go in the next days !

asyates commented:
max_dtt now ;)

note, this is just for mwcs. It is still maxdt for wavelet. Both stretching and wavelet also need some work to finish their coda window param choices (i have included some params as placeholders that are currently not implemented).

ThomasLecocq commented:
Sorry for the delay, it was too much for Skience, and I need time to test it :-) Was wondering if it'd be clever now to change the folder on the disk's naming from "01" to "f01", for filters, 'mwcs01' etc ? 
that tree view is a bit scary (although I like the idea of forcing users to use the API... but yay)
![image](https://github.com/user-attachments/assets/a4cf28c0-0314-4c0a-b02a-fba19f592d45)

asyates commented:
Agree re. improving the folder names. Can look to make this change next week (this week fully booked). Don't hesitate ofc if anything else comes to mind in mean time.
